src/cpu/base.cc

```diff
@@ -1114,6 +1114,232 @@ BaseCPU::diffWithNEMU(ThreadID tid, InstSeqNum seq)
+                    "%#lx\033[0m, GEM5 value: \033[31m%#lx\033[0m\n",
+                    "vsepc", ref_val, gem5_val);
+            diffInfo.errorCsrsValue[CsrRegIndex::vsepc] = 1;
+            diffAllStates->gem5RegFile.vsepc= gem5_val;
+            if (!diff_at)
+                diff_at = ValueDiff;
+        }
+        //vscause
+        gem5_val = readMiscReg(RiscvISA::MiscRegIndex::MISCREG_VSCAUSE, tid);
+        ref_val = diffAllStates->referenceRegFile.vscause;
+        if (gem5_val != ref_val) {
+            warn("Inst [sn:%lli] pc: %#lx\n", seq, diffInfo.pc->instAddr());
+            warn("Diff at \033[31m%s\033[0m Ref value: \033[31m"
+                    "%#lx\033[0m, GEM5 value: \033[31m%#lx\033[0m\n",
+                    "vscause", ref_val, gem5_val);
+            diffInfo.errorCsrsValue[CsrRegIndex::vscause] = 1;
+            diffAllStates->gem5RegFile.vscause= gem5_val;
+            if (!diff_at)
+                diff_at = ValueDiff;
+        }
+        //vstval
+        gem5_val = readMiscReg(RiscvISA::MiscRegIndex::MISCREG_VSTVAL, tid);
+        ref_val = diffAllStates->referenceRegFile.vstval;
+        if (gem5_val != ref_val) {
+            warn("Inst [sn:%lli] pc: %#lx\n", seq, diffInfo.pc->instAddr());
+            warn("Diff at \033[31m%s\033[0m Ref value: \033[31m"
+                    "%#lx\033[0m, GEM5 value: \033[31m%#lx\033[0m\n",
+                    "vstval", ref_val, gem5_val);
+            diffInfo.errorCsrsValue[CsrRegIndex::vstval] = 1;
+            diffAllStates->gem5RegFile.vstval = gem5_val;
+            if (!diff_at)
+                diff_at = ValueDiff;
+        }
+
+        //vsatp
+        gem5_val = readMiscReg(RiscvISA::MiscRegIndex::MISCREG_VSATP, tid);
+        ref_val = diffAllStates->referenceRegFile.vsatp;
+        if (gem5_val != ref_val) {
+            warn("Inst [sn:%lli] pc: %#lx\n", seq, diffInfo.pc->instAddr());
+            warn("Diff at \033[31m%s\033[0m Ref value: \033[31m"
+                    "%#lx\033[0m, GEM5 value: \033[31m%#lx\033[0m\n",
+                    "vsatp", ref_val, gem5_val);
+            diffInfo.errorCsrsValue[CsrRegIndex::vsatp] = 1;
+            diffAllStates->gem5RegFile.vsatp = gem5_val;
+            if (!diff_at)
+                diff_at = ValueDiff;
+        }
+
+        //vsscratch
+        gem5_val = readMiscReg(RiscvISA::MiscRegIndex::MISCREG_VSSCRATCH, tid);
+        ref_val = diffAllStates->referenceRegFile.vsscratch;
+        if (gem5_val != ref_val) {
+            warn("Inst [sn:%lli] pc: %#lx\n", seq, diffInfo.pc->instAddr());
+            warn("Diff at \033[31m%s\033[0m Ref value: \033[31m"
+                    "%#lx\033[0m, GEM5 value: \033[31m%#lx\033[0m\n",
+                    "vsscratch", ref_val, gem5_val);
+            diffInfo.errorCsrsValue[CsrRegIndex::vsscratch] = 1;
+            diffAllStates->gem5RegFile.vsscratch = gem5_val;
+            if (!diff_at)
+                diff_at = ValueDiff;
+        }
```

Comment:
Can we rewrite this chunk of code with a function, a template function, or a macro?

It looks too redundant